src/app/Shared/Services/Api.service.tsx

```diff
@@ -1154,6 +1174,64 @@ export class ApiService {
     );
   }
 
+  checkCredentialForTarget(
+    target: Target,
+    credentials: { username: string; password: string }
+  ): Observable<
+    | {
+        error: Error;
+        errorType: string;
+      }
+    | undefined
+  > {
+    const headers = new Headers();
+    headers.set('X-JMX-Authorization', `Basic ${Base64.encode(`${credentials.username}:${credentials.password}`)}`);
+    return this.graphql<any>(
+      `
+      query CheckCredentialForTarget($filter: TargetNodesFilterInput) {
+        targetNodes(filter: $filter) {
+          name
+          mbeanMetrics {
+            runtime {
+              inputArguments
+            }
+          }
+        }
+      }
+      `,
+      {
+        filter: { name: target.connectUrl },
```

Comment:
Hmm. Seems like I probably made a mistake designing this request filter. The `labels` and `annotations` fields can use syntax like `myLabel in (a, b, c)`, which is equivalent to a syntax like `myLabel == "a" || myLabel == "b" || myLabel == "c"`. It looks like it would've been useful to be able to apply this kind of syntax to other things (node names) than just labels and annotations. I didn't originally envision GraphQL queries being used this way but it does make sense. If we could make requests like this then it would mean the whole test could be done with one request-response pair between client and server, instead of one pair for each target under test. Right now the overall test looks like:

```
client -> server -> targetA
client <- server <__/
...
client -> server -> targetB
client <- server <__/
...
client -> server -> targetC
client <- server <__/
```

But with this properly constructed batch request from the client, it could look like:

```
                            /--> targetA \___
client -> server ---> targetB -----  > server -> client
                           \__> targetC __/
```

In the end there may not be a ton of difference if the client is able to make concurrent requests to the server in the first case, since in the end either way ends up with multiple connections opened to targets at a time, so the overall wall time from first request start to last response end is probably not too much different. But, there is a different in overhead of network traffic and request parsing etc. that can be reduced.

Reply:
https://github.com/cryostatio/cryostat/blob/9bcda3f42ad05349a38d3767f33f11042aa767d2/src/main/resources/types.graphqls#L19

https://github.com/cryostatio/cryostat/blob/9bcda3f42ad05349a38d3767f33f11042aa767d2/src/main/java/io/cryostat/net/web/http/api/v2/graph/FilterInput.java#L69

https://github.com/cryostatio/cryostat/blob/9bcda3f42ad05349a38d3767f33f11042aa767d2/src/main/java/io/cryostat/net/web/http/api/v2/graph/TargetNodesFetcher.java#L108 / https://github.com/cryostatio/cryostat/blob/9bcda3f42ad05349a38d3767f33f11042aa767d2/src/main/java/io/cryostat/net/web/http/api/v2/graph/TargetNodesFetcher.java#L115

https://github.com/cryostatio/cryostat/blob/9bcda3f42ad05349a38d3767f33f11042aa767d2/src/main/java/io/cryostat/net/web/http/api/v2/graph/ActiveRecordingsFetcher.java#L97 / https://github.com/cryostatio/cryostat/blob/9bcda3f42ad05349a38d3767f33f11042aa767d2/src/main/java/io/cryostat/net/web/http/api/v2/graph/ActiveRecordingsFetcher.java#L104

I think this functionality can be retrofitted into the existing system pretty cleanly actually:

```diff
diff --git a/src/main/java/io/cryostat/net/web/http/api/v2/graph/FilterInput.java b/src/main/java/io/cryostat/net/web/http/api/v2/graph/FilterInput.java
index ee783a7c..a845453d 100644
--- a/src/main/java/io/cryostat/net/web/http/api/v2/graph/FilterInput.java
+++ b/src/main/java/io/cryostat/net/web/http/api/v2/graph/FilterInput.java
@@ -67,6 +67,7 @@ class FilterInput {
     enum Key {
         ID("id"),
         NAME("name"),
+        NAMES("names"),
         LABELS("labels"),
         ANNOTATIONS("annotations"),
         SOURCE_TARGET("sourceTarget"),
diff --git a/src/main/java/io/cryostat/net/web/http/api/v2/graph/TargetNodesFetcher.java b/src/main/java/io/cryostat/net/web/http/api/v2/graph/TargetNodesFetcher.java
index 2e6df720..a966d1bd 100644
--- a/src/main/java/io/cryostat/net/web/http/api/v2/graph/TargetNodesFetcher.java
+++ b/src/main/java/io/cryostat/net/web/http/api/v2/graph/TargetNodesFetcher.java
@@ -112,6 +112,16 @@ class TargetNodesFetcher extends AbstractPermissionedDataFetcher<List<TargetNode
                             .filter(n -> Objects.equals(n.getName(), nodeName))
                             .collect(Collectors.toList());
         }
+        if (filter.contains(FilterInput.Key.NAMES)) {
+            List<String> names = filter.get(FilterInput.Key.NAMES);
+            for (String name : names) {
+                result =
+                        result.stream()
+                                .filter(n -> LabelSelectorMatcher.parse(name).test(Map.of("name",
+                                                n.getName())))
+                                .collect(Collectors.toList());
+            }
+        }
         if (filter.contains(FilterInput.Key.LABELS)) {
             List<String> labels = filter.get(FilterInput.Key.LABELS);
             for (String label : labels) {
diff --git a/src/main/resources/types.graphqls b/src/main/resources/types.graphqls
index 410efb66..5b15af3d 100644
--- a/src/main/resources/types.graphqls
+++ b/src/main/resources/types.graphqls
@@ -17,6 +17,7 @@ input EnvironmentNodeFilterInput {
 input TargetNodesFilterInput {
     id: Int
     name: String
+    names: [String]
     labels: [String]
     annotations: [String]
 }
```

haven't tested this, just an idea. This would just allow for a single query to match a given node if its name is not just equal to a specific value, but if its name is a member of a given set. Then clients can make much more flexible requests targeting sets of nodes and doing further queries on those all server-side. Thoughts? If this makes sense I will go ahead and work on a PR.

Reply:
Great idea and makes sense to me! This affects only the `Test All` functionality right? Individual tests will still send individual requests?

Reply:
Right, this would mostly just be useful for `Test All`. The individual testing requests could still use the same query syntax if it's easier to reuse code that way on the frontend though. Basically the difference would be something like:

```graphql
query {
  // $someNode should be passed as a variable, not embedded in the query
  targetNodes(filter: { name: "$someNode" } ) {
    name
  }
}
```

for testing using the single-node specific query

vs

```graphql
query {
  // $someNode list contents should be passed as a variable, not embedded in the query
  targetNodes(filter: { names: ["name in ($someNode)"] } ) {
    name
  }
}
```

for testing a single node with the set matching query.